### PR TITLE
ARROW-17481: [C++][Python] Major performance improvements to CSV reading from S3

### PR DESCRIPTION
**Pull request by marsupialtail**
I messed up the Git rebase again: https://github.com/apache/arrow/pull/13931

So let's now look at this PR instead.

### CODE REVIEW

File: cpp/src/arrow/csv/reader.cc [hunk 1]
```diff
@@ -887,30 +872,37 @@ class StreamingReaderImpl : public ReaderMixin,
...
+        std::move(buffer_generator), MakeChunker(ReaderMixin<InputType>::parse_options_),
+        std::move(after_header),
+        ReaderMixin<InputType>::read_options_.skip_rows_after_names);
     auto parsed_block_gen =
-        MakeMappedGenerator(std::move(block_gen), std::move(parser_op));
-    auto rb_gen = MakeMappedGenerator(std::move(parsed_block_gen), std::move(decoder_op));
+        MakeApplyGenerator(std::move(block_gen), std::move(parser_op), cpu_executor);
+    auto preparse_gen =
+        MakeSerialReadaheadGenerator(parsed_block_gen, cpu_executor->GetCapacity());
+    auto rb_gen = MakeMappedGenerator(std::move(preparse_gen), std::move(decoder_op));
```
Comment (westonpace) [inline cpp/src/arrow/csv/reader.cc:903]:
```suggestion
    auto parsed_ra_gen =
        MakeSerialReadaheadGenerator(parsed_block_gen, cpu_executor->GetCapacity());
    auto rb_gen = MakeMappedGenerator(std::move(parsed_ra_gen), std::move(decoder_op));
```

`preparse_gen` is a bit of an odd name as this is after parsing.

File: cpp/src/arrow/csv/reader.h [hunk 2]
```diff
@@ -25,6 +25,7 @@
 #include "arrow/result.h"
 #include "arrow/type.h"
 #include "arrow/type_fwd.h"
+#include "arrow/util/async_generator.h"
```
Comment (westonpace) [inline cpp/src/arrow/csv/reader.h:28]:
```suggestion
```

I don't think this is needed?  We have been mostly trying to treat `async_generator.h` as an internal header as it includes quite a few different things and can bloat header sizes.

Can you put this in `reader.cc` if needed?

File: cpp/src/arrow/dataset/file_csv.cc [hunk 3]
```diff
@@ -18,6 +18,7 @@
 #include "arrow/dataset/file_csv.h"
 
 #include <algorithm>
+#include <iostream>
```
Comment (westonpace) [inline cpp/src/arrow/dataset/file_csv.cc:21]:
```suggestion
```

**Review by westonpace — CHANGES_REQUESTED** [review]
I scanned through again and I think my approval still stands (though I have a few small changes I'd like to see)

@pitrou did you want to take another look?

**Comment by pitrou** [comment]
I'd rather postpone any complexity-inducing improvements after 10.0.0 is released.

**Comment by marsupialtail** [comment]
This is unfortunate as I would really like to see this PR reflected in 10.0 because I have projects that depend on this. But I suppose it is up to you.

**Comment by pitrou** [comment]
I understand. But we're having stability issues and it's not the right time to potentially introduce others...

**Comment by marsupialtail** [comment]
@pitrou would you be able to look at this now that 10.0 is released?

**Comment by pitrou** [comment]
@marsupialtail I'll do it soon.

**Comment by pitrou** [comment]
I didn't look at this as promised.
However, I also noticed a potential problem with the current generator usage in the CSV reader that needs to be investigated: https://github.com/apache/arrow/issues/14792

**Comment by westonpace** [comment]
> However, I also noticed a potential problem with the current generator usage in the CSV reader that needs to be investigated: https://github.com/apache/arrow/issues/14792

I think we're ok here because we aren't actually consuming that generator async-reentrantly.  The apply generate can be called re-entrantly (which would be a problem) but it appears this PR is using MakeSerialReadaheadGenerator.  I seem to recall we ran into a bug when we tried MakeReadaheadGenerator and I wonder if this was it.

So the current implementation will do parallel-I/O, which is nice, and will interleave parsing and decoding.  However, it does not do parallel parsing.  I think we eventually want this but we will want to find a better way of handling #14792 .  Given that the parallel I/O is already giving some nice benefit when reading from S3 perhaps the parallel parse could be left for a follow-up PR.

**Comment by amol-** [comment]
Closing because it has been untouched for a while, in case it's still relevant feel free to reopen and move it forward 👍

**Pull request closed without merging**